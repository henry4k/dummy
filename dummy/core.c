#include <stdlib.h> // malloc, free
#include <string.h> // memset, strncpy
#include <stdio.h>  // dummyFormat etc
#include <stdarg.h> // dummyFormat etc
#include <stdbool.h>
#include <assert.h>

#include "core.h"


enum
{
    DUMMY_MAX_MESSAGE_LENGTH = 127,
    DUMMY_MAX_TESTS = 32,
    DUMMY_MAX_CLEANUPS = 32,
    DUMMY_INVALID_TEST_INDEX = -1,
    DUMMY_SANDBOX_SKIPPED = -1
};

typedef enum
{
    DUMMY_INITIALIZING,
    DUMMY_RUNNING,
    DUMMY_COMPLETED
} dummyStatus;

typedef enum
{
    DUMMY_TEST_UNDEFINED,
    DUMMY_TEST_STARTING,
    DUMMY_TEST_RUNNING,
    DUMMY_TEST_COMPLETED
} dummyTestStatus;

/**
 * Is called after the test has been completed -
 * regardless of whether it failed or succeeded.
 */
typedef struct
{
    dummyCleanupFunction fn;
    void* data;
} dummyCleanup;

typedef struct
{
    char name[DUMMY_MAX_MESSAGE_LENGTH];
    dummySandbox sandbox;
    dummySandboxableFunction fn;

    dummyTestStatus status;

    dummyTestResult result;
    char abortReason[DUMMY_MAX_MESSAGE_LENGTH];

    bool markedAsTodo;
    char todoReason[DUMMY_MAX_MESSAGE_LENGTH];

    dummyCleanup cleanupStack[DUMMY_MAX_CLEANUPS];
    int cleanupStackSize;
} dummyTest;

typedef struct
{
    dummyStatus status;
    const dummyReporter* reporter;

    dummyTest tests[DUMMY_MAX_TESTS];
    int testCount;
    int currentTestIndex;
} dummyContext;

typedef struct
{
    dummyAbortHandler handler;
    void* context;
} dummyAbortHandlerInfo;


static dummyContext* dummyCurrentContext = NULL;
static dummyAbortHandlerInfo AbortStack[DUMMY_MAX_SANDBOX_DEPTH];
static int AbortStackSize = 0;


bool dummyRunTest( int index );
dummyTest* dummyGetCurrentTest();

void dummyInit( const dummyReporter* reporter )
{
    assert(dummyCurrentContext == NULL);
    dummyCurrentContext = (dummyContext*)malloc(sizeof(dummyContext));
    memset(dummyCurrentContext, 0, sizeof(dummyContext));

    dummyCurrentContext->status = DUMMY_INITIALIZING;

    assert(reporter);
    assert(reporter->began);
    assert(reporter->completed);
    assert(reporter->beganTest);
    assert(reporter->completedTest);
    assert(reporter->log);
    dummyCurrentContext->reporter = reporter;

    dummyCurrentContext->currentTestIndex = DUMMY_INVALID_TEST_INDEX;
}

int dummyRunTests()
{
    dummyContext* ctx = dummyCurrentContext;
    assert(ctx);
    assert(ctx->status == DUMMY_INITIALIZING);

    ctx->reporter->began(ctx->reporter->context);

    ctx->status = DUMMY_RUNNING;
    int failedTests = 0;
    int i = 0;
    for(; i < ctx->testCount; i++)
        if(dummyRunTest(i) == false)
            failedTests++;

    ctx->status = DUMMY_COMPLETED;
    ctx->reporter->completed(ctx->reporter->context);

    assert(ctx->status == DUMMY_COMPLETED);
    free(ctx);
    dummyCurrentContext = NULL;

    return failedTests;
}

void dummyAddTest( const char* name, dummySandbox sandbox, dummySandboxableFunction fn )
{
    dummyContext* ctx = dummyCurrentContext;
    assert(ctx);
    assert(ctx->status == DUMMY_INITIALIZING);
    assert(ctx->testCount < DUMMY_MAX_TESTS);

    dummyTest* test = &ctx->tests[ctx->testCount];
    memset(test, 0, sizeof(dummyTest));

    strncpy(test->name, name, DUMMY_MAX_MESSAGE_LENGTH);
    test->sandbox = sandbox;
    test->fn = fn;

    ctx->testCount++;
}

bool dummyRunTest( int index )
{
    dummyContext* ctx = dummyCurrentContext;
    assert(ctx);
    assert(ctx->status == DUMMY_RUNNING);

    assert(index >= 0);
    assert(index < DUMMY_MAX_TESTS);
    dummyTest* test = &ctx->tests[index];

    // prepare
    ctx->currentTestIndex = index;
    test->status = DUMMY_TEST_STARTING;
    test->result = DUMMY_TEST_PASSED;
    ctx->reporter->beganTest(ctx->reporter->context);

    // run
    test->status = DUMMY_TEST_RUNNING;
    const char* abortReason = NULL;
    const int errorCode = test->sandbox(test->fn, &abortReason);
    switch(errorCode)
    {
        case DUMMY_SANDBOX_SUCCEEDED:
            test->result = DUMMY_TEST_PASSED;
            break;

        case DUMMY_SANDBOX_SKIPPED:
            test->result = DUMMY_TEST_SKIPPED;
            break;

        case DUMMY_SANDBOX_GENERIC_ERROR:
        default:
            test->result = DUMMY_TEST_FAILED;
    }
    if(abortReason)
        strncpy(test->abortReason, abortReason, DUMMY_MAX_MESSAGE_LENGTH);
    test->status = DUMMY_TEST_COMPLETED;

    for(int i = 0; i < test->cleanupStackSize; i++)
    {
        dummyCleanup* cleanup = &test->cleanupStack[i];
        cleanup->fn(cleanup->data);
    }
    test->cleanupStackSize = 0;

    // cleanup
    ctx->reporter->completedTest(ctx->reporter->context);
    ctx->currentTestIndex = DUMMY_INVALID_TEST_INDEX;
    test->status = DUMMY_TEST_UNDEFINED;

    return test->result != DUMMY_TEST_FAILED;
}

int dummyGetTestCount()
{
    const dummyContext* ctx = dummyCurrentContext;
    assert(ctx);

    return ctx->testCount;
}

int dummyGetTestNumber()
{
    const dummyContext* ctx = dummyCurrentContext;
    assert(ctx);

    assert(ctx->currentTestIndex >= 0);
    assert(ctx->currentTestIndex < DUMMY_MAX_TESTS);

    return ctx->currentTestIndex;
}

dummyTest* dummyGetCurrentTest()
{
    dummyContext* ctx = dummyCurrentContext;
    assert(ctx);

    assert(ctx->currentTestIndex >= 0);
    assert(ctx->currentTestIndex < DUMMY_MAX_TESTS);
    dummyTest* test = &ctx->tests[ctx->currentTestIndex];

    return test;
}

const char* dummyGetTestName()
{
    const dummyTest* test = dummyGetCurrentTest();
    assert(test->name);
    assert(strlen(test->name) < DUMMY_MAX_MESSAGE_LENGTH);
    return test->name;
}

dummyTestResult dummyGetTestResult()
{
    const dummyTest* test = dummyGetCurrentTest();
    assert(test->status == DUMMY_TEST_COMPLETED);
    return test->result;
}

const char* dummyGetTestAbortReason()
{
    const dummyTest* test = dummyGetCurrentTest();
    if(test->status == DUMMY_TEST_COMPLETED)
    {
        assert(test->abortReason);
        const int length = strlen(test->abortReason);
        assert(length < DUMMY_MAX_MESSAGE_LENGTH);
        if(length > 0)
            return test->abortReason;
    }

    return NULL;
}

void dummyAddCleanup( dummyCleanupFunction fn, void* data )
{
    dummyTest* test = dummyGetCurrentTest();

    assert(test->cleanupStackSize < DUMMY_MAX_CLEANUPS);
    dummyCleanup* cleanup = &test->cleanupStack[test->cleanupStackSize];

    cleanup->fn = fn;
    cleanup->data = data;

    test->cleanupStackSize++;
}

int dummyTestIsMarkedAsTodo()
{
    return dummyGetCurrentTest()->markedAsTodo;
}

const char* dummyGetTestTodoReason()
{
    const dummyTest* test = dummyGetCurrentTest();

    if(test->markedAsTodo)
    {
        assert(test->todoReason);
        const int length = strlen(test->todoReason);
        assert(length < DUMMY_MAX_MESSAGE_LENGTH);
        if(length > 0)
            return test->todoReason;
    }

    return NULL;
}

const char* dummyFormatV( const char* format, va_list args )
{
    static char buffer[DUMMY_MAX_MESSAGE_LENGTH];
    vsprintf(buffer, format, args);
    return buffer;
}

void dummyAbortTest( dummyTestAbortType type, const char* reason, ... )
{
    int errorCode = DUMMY_SANDBOX_GENERIC_ERROR;
    switch(type)
    {
        case DUMMY_FAIL_TEST:
            errorCode = DUMMY_SANDBOX_GENERIC_ERROR;
            break;

        case DUMMY_SKIP_TEST:
            errorCode = DUMMY_SANDBOX_SKIPPED;
            break;

        default:
            assert(!"Unknown abort type");
    }

    const char* formattedReason = NULL;
    if(reason)
    {
        va_list args;
        va_start(args, reason);
        formattedReason = dummyFormatV(reason, args);
        va_end(args);
    }
    else
    {
        assert(!"dummyAbortTest needs a reason!");
    }

    dummyAbortSandbox(errorCode, formattedReason);
}

void dummyMarkTestAsTodo( const char* reason, ... )
{
    dummyTest* test = dummyGetCurrentTest();
    test->markedAsTodo = true;
    if(reason)
    {
        va_list args;
        va_start(args, reason);
        strncpy(test->todoReason, dummyFormatV(reason, args), DUMMY_MAX_MESSAGE_LENGTH);
        va_end(args);
    }
}

void dummyLog( const char* message, ... )
{
    const dummyContext* ctx = dummyCurrentContext;
    assert(ctx);

    va_list args;
    va_start(args, message);
    const char* formattedMessage = dummyFormatV(message, args);
    va_end(args);

    ctx->reporter->log(ctx->reporter->context, formattedMessage);
}

void dummyPushAbortHandler( dummyAbortHandler handler, void* context )
{
    assert(AbortStackSize <= DUMMY_MAX_SANDBOX_DEPTH);
    AbortStackSize++;
    dummyAbortHandlerInfo* info = &AbortStack[AbortStackSize-1];

    info->handler = handler;
    info->context = context;
}

void dummyPopAbortHandler()
{
    assert(AbortStackSize >= 1);
    AbortStackSize--;
}

void dummyAbortSandbox( int errorCode, const char* reason )
{
    assert(AbortStackSize >= 1);
    dummyAbortHandlerInfo* info = &AbortStack[AbortStackSize-1];

    info->handler(info->context, errorCode, reason);
}
