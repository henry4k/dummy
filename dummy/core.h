#ifndef __DUMMY_CORE_H__
#define __DUMMY_CORE_H__

#ifdef __cplusplus
extern "C"
{
#endif

enum
{
    DUMMY_UNKNOWN_TEST_COUNT = -1
};

typedef enum
{
    DUMMY_TEST_PASSED,
    DUMMY_TEST_FAILED,
    DUMMY_TEST_SKIPPED
} dummyTestResult;

typedef enum
{
    DUMMY_FAIL_TEST,
    DUMMY_SKIP_TEST
} dummyTestAbortType;

enum
{
    DUMMY_SANDBOX_SUCEEDED = 0,
    DUMMY_SANDBOX_GENERIC_ERROR = 1
};

/**
 * Function that is called for a test.
 */
typedef void (*dummyTestFunction)();

typedef void (*dummyCleanupFunction)( void* data );

typedef struct
{
    void* context;

    /**
     * Runs the given function and catches errors.
     *
     * @param abortReason
     * If the protected call is aborted with a reason,
     * the string pointer is passed here.
     *
     * @return
     * Code that classifies the error.
     * If the function ras successfully it returns DUMMY_PROTECTED_CALL_SUCEEDED.
     * Custom error codes may be passed using dummySandbox::abort().
     */
    int (*run)( void* context, dummyTestFunction fn, const char** abortReason );

    /**
     * Aborts the current test.
     *
     * @param reason
     * May be `NULL`.
     *
     * @return
     * Doesn't return.
     */
    void (*abort)( void* context, int errorCode, const char* reason );
} dummySandbox;

typedef struct
{
    void* context;

    void (*began)( void* context );
    void (*completed)( void* context );

    void (*beganTest)( void* context );
    void (*completedTest)( void* context );

    void (*log)( void* context, const char* message );
} dummyReporter;


/**
 * Initializes the test context.
 *
 * This needs to be called before any calls to other test functions.
 * The initial context status is `DUMMY_INITIALIZING`.
 *
 * @param reporter
 * Reporter which will be used by the created context.
 */
void dummyInit( const dummySandbox* sandbox, const dummyReporter* reporter );

/**
 * Runs all added tests and destroys the current context.
 *
 * @return
 * Number of test tests, that failed.
 */
int dummyRunTests();

/**
 * Adds a test to the current context.
 *
 * @param name
 * May be used by the reporter to describe the test.
 *
 * @param fn
 * Is called when the test is being run.
 * See #dummyRunTests
 */
void dummyAddTest( const char* name, dummyTestFunction fn );

/**
 * Count of tests added to the current context.
 */
int dummyGetTestCount();

/**
 * Name of the active test in the current context.
 */
const char* dummyGetTestName();

/**
 * Id of the active test in the current context.
 */
int dummyGetTestNumber();

/**
 * TODO
 */
dummyTestResult dummyGetTestResult();

/**
 * Message describing why the current test aborted
 * or `NULL` if the test hasn't failed (yet).
 */
const char* dummyGetTestAbortReason();

/**
 * Adds a cleanup function to the current test.
 *
 * @param data
 * Data pointer that is passed to the cleanup function.
 */
void dummyAddCleanup( dummyCleanupFunction fn, void* data );

int dummyTestIsMarkedAsTodo();

/**
 * If set, it returns the reason why the current test has been marked as TODO.
 * May be `NULL` if no reason has been given.
 */
const char* dummyGetTestTodoReason();

/**
 * Aborts current test with the given reason.
 *
 * @param reason
 * Explains why the test was aborted -
 * this information is available to the reporter.
 *
 * @return
 * Doesn't return.
 */
void dummyAbortTest( dummyTestAbortType type, const char* reason, ... );

/**
 * Marks current test as TODO.
 *
 * @param reason
 * Explains why the test has been marked as TODO -
 * this information is available to the reporter.
 * May be `NULL` to signal an unknown reason.
 *
 * @note
 * Calling this function more than once in a test,
 * may overwrite the previous TODO reason.
 */
void dummyMarkTestAsTodo( const char* reason, ... );


/**
 * Reports a diagnostic message.
 */
void dummyLog( const char* message, ... );


#ifdef __cplusplus
}
#endif

#endif
