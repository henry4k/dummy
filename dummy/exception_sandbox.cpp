#include <stddef.h> // NULL
#include <exception>

#include "signal.h"
#include "exception_sandbox.h"


class AbortError
{
public:
    int m_ErrorCode;
    const char* m_Reason;

    explicit AbortError( int errorCode, const char* reason ) :
        m_ErrorCode(errorCode),
        m_Reason(reason)
    {
    }
};

static void abortTest( void* context_, int errorCode, const char* reason )
{
    throw AbortError(errorCode, reason);
}

static void signalHandler( int signal )
{
    throw AbortError(DUMMY_SANDBOX_GENERIC_ERROR, dummySignalToAbortReason(signal));
}

static int runTest( void* context_, dummyTestFunction fn, const char** abortReason )
{
    const char* reason = NULL;
    int exitCode = DUMMY_SANDBOX_SUCEEDED;
    dummySetSignals(signalHandler);

    try
    {
        fn();
    }
    catch( AbortError& e )
    {
        reason = e.m_Reason;
        exitCode = e.m_ErrorCode;
    }
    catch( std::exception& e )
    {
        reason = e.what();
        exitCode = DUMMY_SANDBOX_GENERIC_ERROR;
    }
    catch( ... )
    {
        reason = "An unknown exception has been thrown.";
        exitCode = DUMMY_SANDBOX_GENERIC_ERROR;
    }

    dummySetSignals(SIG_DFL);
    if(abortReason)
        *abortReason = reason;
    return exitCode;
}

const dummySandbox* dummyGetExceptionSandbox()
{
    static dummySandbox sandbox;

    sandbox.context = NULL;
    sandbox.run = runTest;
    sandbox.abort = abortTest;

    return &sandbox;
}
