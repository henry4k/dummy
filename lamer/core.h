#ifndef __LAMER_CORE_H__
#define __LAMER_CORE_H__

#ifdef __cplusplus
extern "C"
{
#endif

enum
{
    LAMER_UNKNOWN_TEST_COUNT = -1
};

typedef enum
{
    LAMER_INITIALIZING,
    LAMER_RUNNING,
    LAMER_COMPLETED
} lamerStatus;

typedef enum
{
    LAMER_TEST_UNDEFINED,
    LAMER_TEST_STARTING,
    LAMER_TEST_RUNNING,
    LAMER_TEST_COMPLETED
} lamerTestStatus;

typedef enum
{
    LAMER_TEST_PASSED,
    LAMER_TEST_FAILED,
    LAMER_TEST_SKIPPED
} lamerTestResult;

typedef enum
{
    LAMER_FAIL_TEST,
    LAMER_SKIP_TEST
} lamerTestAbortType;

typedef struct
{
    void* context;

    void (*began)( void* context );
    void (*completed)( void* context );

    void (*beganTest)( void* context );
    void (*completedTest)( void* context );

    void (*diag)( void* context, const char* message );
} lamerReporter;


/**
 * Initializes the test context.
 *
 * This needs to be called before any calls to other test functions.
 * The initial context status is `LAMER_INITIALIZING`.
 *
 * @param reporter
 * Reporter which will be used by the created context.
 */
void lamerInit( const lamerReporter* reporter );

/**
 * Runs all added tests and destroys the current context.
 *
 * @return
 * Number of test tests, that failed.
 */
int lamerRunTests();

/**
 * Function that is called for a test.
 */
typedef void (*lamerTestFunction)();

/**
 * Adds a test to the current context.
 *
 * @param name
 * May be used by the reporter to describe the test.
 *
 * @param fn
 * Is called when the test is being run.
 * See #lamerRunTests
 *
 * @return
 * Id of the test.
 */
int lamerAddTest( const char* name, lamerTestFunction fn );

/**
 * Status of the current context.
 */
lamerStatus lamerGetStatus();

/**
 * Count of tests added to the current context.
 */
int lamerGetTestCount();

/**
 * Name of the active test in the current context.
 */
const char* lamerGetTestName();

/**
 * Id of the active test in the current context.
 */
int lamerGetTestNumber();

/**
 * Status of the active test in the current context.
 */
lamerTestStatus lamerGetTestStatus();

/**
 * TODO
 */
lamerTestResult lamerGetTestResult();

/**
 * Message describing why the current test aborted
 * or `NULL` if the test hasn't failed (yet).
 */
const char* lamerGetTestAbortReason();

int lamerTestIsMarkedAsTodo();

const char* lamerGetTestTodoReason();

/**
 * Aborts current test with the given reason.
 *
 * @param reason
 * Explains why the test was aborted - this information is available
 * to the reporter. May be `NULL` to signal an unknown reason.
 *
 * @return
 * Doesn't return.
 */
void lamerAbortTest( lamerTestAbortType type, const char* reason, ... );

/**
 * Marks current test as TODO.
 *
 * @param reason
 * Explains why the test has been marked as TODO -
 * this information is available to the reporter.
 * May be `NULL` to signal an unknown reason.
 *
 * @note
 * Calling this function more than once in a test,
 * may overwrite the previous TODO reason.
 */
void lamerMarkTestAsTodo( const char* reason, ... );


#ifdef __cplusplus
}
#endif

#endif
