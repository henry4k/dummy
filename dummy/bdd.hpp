#ifndef __LAMER_BDD_HPP__
#define __LAMER_BDD_HPP__

#include <string>
#include "core.h"

namespace lamer
{

class Describe
{
public:
    Describe( const char* subject );
    Describe& it( const char* behaviour, lamerTestFunction fn );

private:
    std::string m_Subject;
};


// ---- implementation ----

Describe::Describe( const char* subject ) :
    m_Subject(subject)
{
}

Describe& Describe::it( const char* behaviour, lamerTestFunction fn )
{
    lamerAddTest((m_Subject+" "+behaviour).c_str(), fn);
    return *this;
}

}

#endif
