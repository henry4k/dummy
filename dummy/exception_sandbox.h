#ifndef __DUMMY_EXCEPTION_SANDBOX_H__
#define __DUMMY_EXCEPTION_SANDBOX_H__

#include "core.h"
#include "config.h"

#if !defined(DUMMY_CXX_MODE)
#warning "DUMMY_CXX_MODE is not defined, error descriptions may not be accurate."
#endif

#ifdef __cplusplus
extern "C"
{
#endif

const dummySandbox* dummyGetExceptionSandbox();

#ifdef __cplusplus
}
#endif

#endif
