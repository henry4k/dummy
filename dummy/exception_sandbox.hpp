#ifndef __DUMMY_EXCEPTION_SANDBOX_HPP__
#define __DUMMY_EXCEPTION_SANDBOX_HPP__

#include "core.h"

/**
 * Run `fn` in a sandbox, that catches exceptions and signals.
 * @see dummySandbox
 */
int dummyExceptionSandbox( dummySandboxableFunction fn, const char** abortReason );

#endif
