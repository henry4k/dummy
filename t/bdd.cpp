#include "../dummy/core.h"
#include "../dummy/max_reporter.h"
#include "../dummy/bdd.hpp"

using namespace dummy;

int main()
{
    dummyInit(dummyGetMaxReporter());

    Describe("fizzbuzzer module")
        .it("does things", [](){
            // try doing things here
        })
        .it("is a slacker", [](){
            dummyMarkTestAsTodo("Implement lazyness");
            // prove that fizzbuzzer is lazy as fuck
        })
        .it("can't handle bloops", [](){
            dummyMarkTestAsTodo("Implement exception");
            // dummy if bloops cause fizzbuzzer to fail
            dummyAbortTest(DUMMY_SKIP_TEST, "Exception not implemented yet");
        });

    Describe("bloop")
        .it("can make a sound", [](){
            // emulate a sound
            dummyAbortTest(DUMMY_FAIL_TEST, "Bloop is silent!");
        })
        .it("looks weird", [](){
            dummyMarkTestAsTodo("Bloop needs to look weirder.");
            // bloops should always look weird
            dummyAbortTest(DUMMY_FAIL_TEST, "Bloop doesn't looks weird enough!");
        });

    dummyRunTests();
}
