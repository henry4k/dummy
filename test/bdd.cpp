#include <dummy/core.h>
#include <dummy/exception_sandbox.h>
#include <dummy/bdd.hpp>

#include "test_reporter.h"


using namespace dummy;

int main()
{
    dummyInit(dummyGetTestReporter());

    Describe("fizzbuzzer module")
        .use(dummyExceptionSandbox)

        .it("does things", [](){
            // try doing things here
        })
        .it("is a slacker", [](){
            dummyMarkTestAsTodo("Implement lazyness");
            // prove that fizzbuzzer is lazy as fuck
        })
        .it("can't handle bloops", [](){
            dummyMarkTestAsTodo("Implement exception");
            // test if bloops cause fizzbuzzer to fail
            dummyAbortTest(DUMMY_SKIP_TEST, "Exception not implemented yet");
        });

    Describe("bloop")
        .use(dummyExceptionSandbox)

        .it("can make a sound", [](){
            // emulate a sound
            dummyAbortTest(DUMMY_FAIL_TEST, "Bloop is silent!");
        })
        .it("looks weird", [](){
            dummyMarkTestAsTodo("Bloop needs to look weirder.");
            // bloops should always look weird
            dummyAbortTest(DUMMY_FAIL_TEST, "Bloop doesn't looks weird enough!");
        });

    dummyRunTests();
}
