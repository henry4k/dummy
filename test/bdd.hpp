#ifndef __TEST_BDD_HPP__
#define __TEST_BDD_HPP__

#include <string>
#include "core.h"

namespace test
{

class Describe
{
public:
    Describe( const char* subject );
    Describe& it( const char* behaviour, testCaseFunction fn );

private:
    std::string m_Subject;
};


// ---- implementation ----

Describe::Describe( const char* subject ) :
    m_Subject(subject)
{
}

Describe& Describe::it( const char* behaviour, testCaseFunction fn )
{
    testAddCase((m_Subject+" "+behaviour).c_str(), fn);
    return *this;
}

}

#endif
