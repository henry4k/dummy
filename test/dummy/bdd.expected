test count = 5
test number = 0
test name = 'fizzbuzzer module does things'
test result = 'passed'
test abort reason = '(null)'
test todo = 0
test todo reason = '(null)'
test done
test number = 1
test name = 'fizzbuzzer module is a slacker'
test result = 'passed'
test abort reason = '(null)'
test todo = 1
test todo reason = 'Implement lazyness'
test done
test number = 2
test name = 'fizzbuzzer module can't handle bloops'
test result = 'skipped'
test abort reason = 'Exception not implemented yet'
test todo = 1
test todo reason = 'Implement exception'
test done
test number = 3
test name = 'bloop can make a sound'
test result = 'failed'
test abort reason = 'Bloop is silent!'
test todo = 0
test todo reason = '(null)'
test done
test number = 4
test name = 'bloop looks weird'
test result = 'failed'
test abort reason = 'Bloop doesn't looks weird enough!'
test todo = 1
test todo reason = 'Bloop needs to look weirder.'
test done
testing done
